geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,582 @@
+fn fix_orientation<T: GeoFloat>(mut poly: Polygon<T>) -> Polygon<T> {
+    poly.exterior_mut(|ls| ls.make_ccw_winding());
+
+    // just commenting this out for the moment. The algorithm only works on exteriors so this
+    // shouldn't be needed
+    //
+    //poly.interiors_mut(|ls| ls.iter_mut().for_each(|ls| ls.make_ccw_winding()));
+
+    poly
+}
+
+/// checks whether the to lines are equal or inverted forms of each other
+fn same_line<T: GeoFloat>(l1: &Line<T>, l2: &Line<T>) -> bool {
+    let flipped_l2 = Line::new(l2.end, l2.start);
+    l1 == l2 || l1 == &flipped_l2
+}
+
+/// given a collection of lines from multiple polygons, this returns all but the shared lines
+fn find_boundary_lines<T: GeoFloat>(lines: Vec<Line<T>>) -> Vec<Line<T>> {
+    let enumerated_lines = || lines.iter().enumerate();
+    enumerated_lines()
+        // only collect lines that don't have a duplicate in the set
+        .filter_map(|(i, line)| {
+            let same_line_exists = enumerated_lines()
+                .filter(|&(j, _)| j != i)
+                .any(|(_, l)| same_line(line, l));
+            (!same_line_exists).then_some(*line)
+        })
+        .collect::<Vec<_>>()
+}
+
+// Notes for future: This probably belongs into a `Validify` trait or something
+/// finds holes in polygon exterior and fixes them
+///
+/// This is important for scenarios like the banana polygon. Which is considered invalid
+/// https://www.postgis.net/workshops/postgis-intro/validity.html#repairing-invalidity
+fn find_and_fix_holes_in_exterior<F: GeoFloat>(mut poly: Polygon<F>) -> Polygon<F> {
+    fn detect_if_rings_closed_with_point<F: GeoFloat>(
+        points: &mut Vec<&Coord<F>>,
+        p: &Coord<F>,
+    ) -> Option<Vec<Coord<F>>> {
+        // early return here if nothing was found
+        let pos = points.iter().position(|&c| c == p)?;
+
+        // create ring by collecting the points if something was found
+        let ring = points
+            .drain(pos..)
+            .chain(std::iter::once(p))
+            .cloned()
+            .collect::<Vec<_>>();
+        Some(ring)
+    }
+
+    // find rings
+    let rings = {
+        let (points, mut rings) =
+            poly.exterior()
+                .into_iter()
+                .fold((vec![], vec![]), |(mut points, mut rings), coord| {
+                    if let Some(ring) = detect_if_rings_closed_with_point(&mut points, coord) {
+                        rings.push(ring);
+                    }
+                    points.push(coord);
+                    (points, rings)
+                });
+
+        // add leftover coords as last ring
+        let last_ring = points.into_iter().cloned().collect::<Vec<_>>();
+        rings.push(last_ring);
+
+        rings
+    };
+
+    // convert to polygons for containment checks
+    let mut rings = rings
+        .into_iter()
+        // filter out degenerate polygons which may be produced from the code above
+        .filter(|cs| cs.len() >= 3)
```

Comment:
Do you have an example of what kind of inputs would hit this case? I'm wondering if this kind of "validification" is  losing important information.

Reply:
I think we ran into this when we got invalid polygons as inputs. With this, the algo even worked for the invalid polygons we got.